### PR TITLE
Use second rate limit bucket for code completions

### PR DESCRIPTION
**Pull request by mrnugget**
This basically duplicates the existing code for completion rate limits and adds a second rate limit for _code_ completions.

What's in here:

- adds new column to `users` database table & database store methods
- adds new config value to site config so we can set global code-completions-per-day limit
- GraphQL mutations/query to read and update quota per user
- tests
- adds a "scope" to the rate limiters that we use for completion endpoints and then uses one of the two limits depending on which scope is used, i.e. if it's "completions" or "code completions"

What's *NOT* in here:

- changes to the admin UI so that we can see and update the quota, @philipp-spiess will add that in a separate PR

## Test plan

- Manual testing by hooking up Cody for VS Code 0.0.10 to local instance, observing Redis keys/values, seeing error response in VS Code and, of course, printf-debugging

### CODE REVIEW

File: enterprise/internal/completions/streaming/codecompletion.go [hunk 1]
```diff
@@ -70,6 +70,7 @@ func (h *codeCompletionHandler) ServeHTTP(w http.ResponseWriter, r *http.Request
 			respondRateLimited(w, unwrap)
 			return
 		}
+		fmt.Println("woooooops")
```
Comment (eseliger) [inline enterprise/internal/completions/streaming/codecompletion.go:73]: indeed.

Comment (mrnugget) [reply enterprise/internal/completions/streaming/codecompletion.go:73]: :fingerguns:

**Comment by sourcegraph-buildkite** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.00% (+0.27 kb) | 0.00% (+0.27 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/mrn/two-bucket-rate-limit/compare-report.html;tab=live_object) for a full comparison between the commits cf5c8c0 and a3d8f3d or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: enterprise/internal/completions/streaming/limiter.go [hunk 2]
```diff
@@ -130,19 +147,34 @@ func (r *rateLimiter) TryAcquire(ctx context.Context) (err error) {
 	return nil
 }
 
-func userKey(userID int32) string {
+func userKey(userID int32, scope RateLimitScope) string {
+	if scope == RateLimitScopeCodeCompletion {
+		return fmt.Sprintf("user:%d:code_completion_requests", userID)
+	}
```
Comment (eseliger) [inline enterprise/internal/completions/streaming/limiter.go:153]: if we made RateLimitScope string and not int, we wouldn't need so many switches in this code, and could do something like

```suggestion
func userKey(userID int32, scope RateLimitScope) string {
		return fmt.Sprintf("user:%d:%s_requests", userID, scope)
```

File: enterprise/internal/completions/streaming/limiter.go [hunk 3]
```diff
@@ -14,31 +14,47 @@ import (
...
+	RateLimitScopeCodeCompletion
+)
+
+func (s RateLimitScope) String() string {
+	if s == RateLimitScopeCompletion {
+		return "completions"
+	}
+	return "code completions"
+}
+
```
[the remark this answers is not in the document]
Comment (eseliger) [reply enterprise/internal/completions/streaming/limiter.go:30]: This seems fine to me, left one comment below on int vs string enums but not blocker

**Review by eseliger — APPROVED** [review]
give me UI 😢

File: enterprise/internal/completions/streaming/limiter.go (hunk 2 above)
Comment (mrnugget) [reply enterprise/internal/completions/streaming/limiter.go:153]: done!

**Pull request merged**